### PR TITLE
[OG-82] Hall of fame backend

### PR DESCRIPTION
**Pull request by LisieckaK00**
I've created two views. They need to be added manually to the database, that's why I'm adding `hof.sql` file with the code.

Based on requirements: https://github.com/krudny/Polymorphia/pull/32.
- `xpDetails` contains string paris "name":"points", where possible values for "name" are: bonus, total and event section names
- backend returns `studentName`, when the request is made by instructor or coordinator
- default values are provided for all params except `courseId`
- there is no need for a different endpoint for podium, it'll be enough to pass `size=3` as param (`GET /hall-of-fame?courseId=1&size=3`)

TODO before merging:
- I need to fix ranking, as right now for students with same points they'll get same position in ranking (the view will be modified)
- Merge with frontend

TODO after backend changes:
- Right now points are just random values. After backend refactor I'll need to update the logic to count the points (`student_score_detail_view` update)

### CODE REVIEW

**Comment by krudny** [comment]
Thank you for update. 

Probably will do review after it is ready. 

Please note that in that PR you should integrate that with frontend and also our supervisor wants "find me" function. 
You don't have to implement frontend logic for that, please only create backend logic and receive data on frontend in service layer.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/controller/HallOfFameController.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    public Page<HallOfFameRecordDto> get(
+            @RequestParam Long courseId,
+            @RequestParam(defaultValue = "0") int page,
+            @RequestParam(defaultValue = "50") int size,
+            @RequestParam(defaultValue = "") String searchTerm,
+            @RequestParam(defaultValue = "total") String sortBy,
+            @RequestParam(defaultValue = "desc") String sortOrder,
+            @RequestParam(required = false) List<String> groups
+    ) {
+        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/controller/HallOfFameController.java:37]: That will be used in every controller, maybe worth to extract that logic into separate function.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/response/user/StudentDetailsResponseDto.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.agh.polymorphia_backend.dto.response.user;
+
+import lombok.Builder;
+
+@Builder
+public record StudentDetailsResponseDto(
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/response/user/StudentDetailsResponseDto.java:6]: There's issue/task how to handle user data, but having `studentName` there is a security threat for me. Someone might forgot to make it `null` by mistake and that's a serious problem for us. 

RODO policies and AGH regulations.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/model/HallOfFame.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    @Id
+    private Long animalId;
+    private String animalName;
+    private String studentName;
+    private String groupName;
+    private Long courseId;
+    private Integer position;
+    private String evolutionStage;
+    private String imageUrl;
+    private BigDecimal rawXpSum;
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/model/HallOfFame.java:25]: Do we care about flat and percentage bonuses? We might, but on frontend we have only "bonuses" and "total"  fields and you might have to modify that.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        Map<Long, Map<String, String>> detailsByAnimalId = groupScoreDetails(animalIds);
+        List<HallOfFameRecordDto> dtoList = getDtoList(pageResult.stream(), detailsByAnimalId, includeStudentName);
+
+        return new PageImpl<>(dtoList, pageable, pageResult.getTotalElements());
+    }
+
+    private Map<Long, Map<String, String>> groupScoreDetails(List<Long> animalIds) {
+        List<StudentScoreDetail> details = scoreDetailRepository.findByAnimalIdIn(animalIds);
+        Map<Long, Map<String, String>> result = new HashMap<>();
+        for (StudentScoreDetail d : details) {
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java:79]: Not a fan of one letter variables, even temporary. `ssd : details` is also not a valid solution :)

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+        Map<Long, Map<String, String>> detailsByAnimalId = groupScoreDetails(animalIds);
+        List<HallOfFameRecordDto> dtoList = getDtoList(pageResult.stream(), detailsByAnimalId, includeStudentName);
+
+        return new PageImpl<>(dtoList, pageable, pageResult.getTotalElements());
+    }
+
+    private Map<Long, Map<String, String>> groupScoreDetails(List<Long> animalIds) {
+        List<StudentScoreDetail> details = scoreDetailRepository.findByAnimalIdIn(animalIds);
+        Map<Long, Map<String, String>> result = new HashMap<>();
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java:78]: maybe some empty line befor `for`?

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    private final StudentScoreDetailRepository scoreDetailRepository;
+    private final HallOfFameRepository hallOfFameRepository;
+    private final HallOfFameMapper hallOfFameMapper;
+
+    private static final Map<String, String> OVERVIEW_FIELDS = Map.ofEntries(Map.entry("animalName", "animalName"), Map.entry("total", "totalXpSum"), Map.entry("bonus", "totalBonusSum"));
+    private static final String NUMBER_FORMAT = "%.02f";
+
+    public Page<HallOfFameRecordDto> getHallOfFame(
+            Long courseId,
+            int page,
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java:32]: There you have new line for each param, below not, we need to standarize that somehow.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            String sortBy,
+            String sortOrder,
+            List<String> groupNames,
+            boolean includeStudentName
+    ) {
+        if (searchTerm == null || searchTerm.isBlank()) {
+            searchTerm = "";
+        }
+        if (!sortOrder.equals("asc") && !sortOrder.equals("desc")) {
+            throw new IllegalArgumentException("Invalid sortOrder: " + sortOrder);
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java:44]: So many ifs, some ternaries, maybe we could simplify that

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+            scoreDetailsList.put("bonus", formatNumber(hallOfFame.getFlatBonusSum().add(hallOfFame.getPercentageBonusSum())));
+        }
+
+        if (!scoreDetailsList.containsKey("total")) {
+            scoreDetailsList.put("total", formatNumber(hallOfFame.getTotalXpSum()));
+        }
+
+        return HallOfFameRecordDto.builder()
+                .userDetails(userDetails)
+                .xpDetails(scoreDetailsList)
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java:36]: Is there a reason it is called `scoreDetailsList`, not `xpDetails`? And also think about renaming everywhere to `pointsDetails` but up to your preferences.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        List<HallOfFameRecordDto> dtoList = getDtoList(hofViews.stream(), detailsByAnimalId, includeStudentName);
+
+        long total = hallOfFameRepository.countByCourseIdAndFilters(courseId, searchTerm, groups);
+
+        return new PageImpl<>(dtoList, PageRequest.of(page, size), total);
+    }
+
+    private List<HallOfFame> getSortedHallOfFame(List<Long> animalIds){
+        List<HallOfFame> hofViews = hallOfFameRepository.findByAnimalIdIn(animalIds);
+
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java:113]: I guess that empty line should be after hashmap, not before

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.Map;
+
+import static com.agh.polymorphia_backend.service.HallOfFameService.formatNumber;
+
+@Service
+@AllArgsConstructor
+public class HallOfFameMapper {
+    public HallOfFameRecordDto hallOfFameToRecordDto(HallOfFame hallOfFame, Map<String, String> scoreDetailsList, boolean includeStudentName) {
+        StudentDetailsResponseDto userDetails = StudentDetailsResponseDto.builder()
+                .studentName(includeStudentName ? hallOfFame.getStudentName() : null)
```
[the remark this answers is not in the document]
Comment (LisieckaK00) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java:17]: It might make sense to check permissions here instead of passing `includeStudentName` from the controller, but I’m not sure if that fits well with the mapper’s responsibility.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java (hunk 7 above)
Comment (LisieckaK00) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java]:
```
getHallOfFame: async (
    page: number,
    size: number,
    searchTerm: string, -> search by name
    sortBy?: string, -> name or category from xpDetails
    sortOrder?: "asc" | "desc",
    groups?: string[]
  )
```
This was the expected behavior. However, currently there’s no distinction whether `searchTerm` refers to the student name or the animal name. For instructors, should I match the term against both? For example, if the `searchTerm` is 'Anna', should the results include a student named Anna as well as any student whose animal is named Anna?

Comment (krudny) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java]: Instructor might have toggle to switch between student and animal name. Backend should handle that. `SortTerm` refers to list instructor sees.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/AuthService.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package com.agh.polymorphia_backend.service;
+
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+@Service
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/AuthService.java:10]: Good

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/controller/HallOfFameController.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+@RestController
+@RequestMapping("/hall-of-fame")
+@AllArgsConstructor
+public class HallOfFameController {
+    private final HallOfFameService hallOfFameService;
+
+    @GetMapping
+    public Page<HallOfFameRecordDto> get(
+            @RequestParam Long courseId,
```
[the remark this answers is not in the document]
Comment (krudny) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/controller/HallOfFameController.java:26]: But who, where and how have to store courseId? 

(i dont know, just asking, even when frontend have to do that, backend engineers should propose/implement solution to that)

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/util/NumberFormatter.java [hunk 13]
```diff
@@ -0,0 +1,14 @@
+package com.agh.polymorphia_backend.util;
+
+import java.math.BigDecimal;
+import java.util.Locale;
+
+public final class NumberFormatter {
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/util/NumberFormatter.java:6]: Good, but use lombok @UtilityClass I guess, not private constructor.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
+package com.agh.polymorphia_backend.service;
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java:1]: Its 11pm, I see that 

image

And I have some panic attack. I dont have enough courage to try to read that. I know Java is not the most readable language but we have to do something with this service. 

image

I was thought that function should have at most three-four params. Now I see 3 LINES of params.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/controller/HallOfFameController.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+            @RequestParam(defaultValue = "") String searchTerm,
+            @RequestParam(defaultValue = "total") String sortBy,
+            @RequestParam(defaultValue = "desc") String sortOrder,
+            @RequestParam(required = false) List<String> groups
+    ) {
+        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+        User user = (User) auth.getPrincipal();
+        Collection<? extends GrantedAuthority> authorities = user.getAuthorities();
+        boolean includeStudentName = authorities.contains(UserType.COORDINATOR) || authorities.contains(UserType.INSTRUCTOR);
+        return hallOfFameService.getHallOfFame(courseId, page, size, searchTerm, sortBy, sortOrder, groups, includeStudentName);
```
[the remark this answers is not in the document]
Comment (LisieckaK00) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/controller/HallOfFameController.java]: created dto

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/response/user/StudentDetailsResponseDto.java (hunk 2 above)
Comment (LisieckaK00) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/response/user/StudentDetailsResponseDto.java]: I created two different dtos, `StudentDetailsWithNameResponseDto` and `StudentDetailsWithoutNameResponseDto`, if you see a better solution let me know

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/model/hall_of_fame/OverviewField.java [hunk 16]
```diff
@@ -0,0 +1,25 @@
+package com.agh.polymorphia_backend.model.hall_of_fame;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+public enum OverviewField {
+    ANIMAL_NAME("animalName", "animalName"),
```
[the remark this answers is not in the document]
Comment (LisieckaK00) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/model/hall_of_fame/OverviewField.java]: I created constuctor with one arg

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java (hunk 14 above)
Comment (LisieckaK00) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/HallOfFameService.java]: I changed the approach. If you don't like it this time too, suggest some specific change or we'll be moving in circles and this PR will never be merged

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (LisieckaK00) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java]: Moved checking authorities to the mapper

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/util/NumberFormatter.java (hunk 13 above)
Comment (LisieckaK00) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/util/NumberFormatter.java:9]: done

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/request/HallOfFameRequestDto.java [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    private long courseId;
+    private int page = 0;
+    private int size = 50;
+    private String searchTerm = "";
+    private String sortBy = "total";
+    private SortOrder sortOrder = SortOrder.DESC;
+    @Setter
+    private List<String> groups;
+
+    @JsonSetter(nulls = Nulls.SKIP)
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/request/HallOfFameRequestDto.java:28]: I dont know java, but i believe we could combine @Setter to class and @NotNull on every field.

I dont understand why we are going to process request with default values when someone is requesting null fields. 

We can leave that, I see some benefits, but please talk with @ustankie @kligarski about standarizing this approach when and why we are processing null fields.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/request/HallOfFameRequestDto.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+    public void setSortOrder(String sortOrder) {
+        this.sortOrder = SortOrder.fromString(sortOrder);
+    }
+
+    public Pageable getPageable() {
+        return PageRequest.of(page, size, Sort.by(sortOrder.getDirection(), sortBy));
+    }
+
+    public String[] getGroupsAsArray(){
+        if (groups == null || groups.isEmpty()){
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/request/HallOfFameRequestDto.java:62]: space after ))

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/response/user/StudentDetailsWithNameResponseDto.java [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+public record StudentDetailsWithNameResponseDto (
+        String studentName,
+        String animalName,
+        String evolutionStage,
+        String group,
+        String imageUrl,
+        int position
+) implements StudentDetailsResponseDto {
+
+    public StudentDetailsWithNameResponseDto(String studentName, StudentDetailsWithoutNameResponseDto studentDetails) {
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/response/user/StudentDetailsWithNameResponseDto.java:12]: In frontend we have UserDetailsDTO, please remember to standarize that.

And think how to handle instructors itself. Do they have animals like students etc.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+        StudentDetailsWithoutNameResponseDto studentDetailsWithoutName = StudentDetailsWithoutNameResponseDto.builder()
+                .animalName(hallOfFame.getAnimalName())
+                .evolutionStage(hallOfFame.getEvolutionStage())
+                .group(hallOfFame.getGroupName())
+                .imageUrl(hallOfFame.getImageUrl())
+                .position(hallOfFame.getPosition())
+                .build();
+
+        boolean includeStudentName = authService.hasAnyRole(List.of(UserType.COORDINATOR, UserType.INSTRUCTOR));
+        StudentDetailsResponseDto userDetails = includeStudentName ?
```
Comment (krudny) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java:32]: User and student in one line, misleading

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+
+        if (!xpDetails.containsKey(OverviewField.BONUS.getKey())) {
+            xpDetails.put(OverviewField.BONUS.getKey(), NumberFormatter.format(hallOfFame.getTotalBonusSum()));
+        }
+
+        if (!xpDetails.containsKey(OverviewField.TOTAL.getKey())) {
+            xpDetails.put(OverviewField.TOTAL.getKey(), NumberFormatter.format(hallOfFame.getTotalXpSum()));
+        }
+
+        return HallOfFameRecordDto.builder()
```
Comment (ustankie) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java:45]: This method does a lot of things, maybe it could be split into some smaller ones?

**Review by ustankie — COMMENTED** [review]
I left some comments, what I also noticed is that most of the methods are quite big and there's a lot happening inside them. I'm not sure if it is possible to reduce them, but maybe it would be good to try. For Dtos creation for example a separate mapper could be used, to obey the SRP rule. 

What I liked is that the search is done in quite generalized way, thanks!

File: polymorphia-backend/src/main/resources/db/migration/R__create_hall_of_fame_views.sql [hunk 22]
```diff
@@ -0,0 +1,66 @@
+DROP VIEW IF EXISTS hall_of_fame_view;
```
[the remark this answers is not in the document]
Comment (LisieckaK00) [reply polymorphia-backend/src/main/resources/db/migration/R__create_hall_of_fame_views.sql:1]: Flyway allows two different prefixes. I used `R` for repeatable migration. This means that whenever this file is changed, the migration will occur. I chose this method as I know this view will be changed (random values right now), and it's easier to test locally, without the need to create many migrations for the same view. If you prefer, I may change it to versioned migration.

This is a view, so not a physical table. As a result, the `xp` and bonuses will always be up-to-date.  If we encounter any issues with performance, it's possible to convert this to materialized view (it'll be saved as a table). However, then it's need to be refreshed manually. That's why I'd start with this option.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/repository/hall_of_fame/HallOfFameRepository.java [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+                        (:#{#requestDto.searchBy().searchByStudent()} = TRUE AND LOWER(h.studentName) LIKE LOWER(CONCAT('%', :#{#requestDto.searchTerm()}, '%')))
+                    ))
+                AND (:#{#requestDto.groups().isEmpty()} = TRUE OR h.groupName IN :#{#requestDto.groups()})
+            """)
+    Page<HallOfFame> findHofPage(
+            @Param("requestDto") HallOfFameRequestDto requestDto,
+            Pageable pageable
+    );
+
+    String WHERE_NATIVE = """
```
Comment (ustankie) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/repository/hall_of_fame/HallOfFameRepository.java:31]: Minor: variables should be defined above methods

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/response/hall_of_fame/HallOfFameRecordDto.java [hunk 24]
```diff
@@ -0,0 +1,12 @@
+package com.agh.polymorphia_backend.dto.response.hall_of_fame;
+
+import com.agh.polymorphia_backend.dto.response.user.StudentDetailsResponseDto;
+import lombok.Builder;
+
+import java.util.Map;
+
+@Builder
+public record HallOfFameRecordDto (
```
Comment (ustankie) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/response/hall_of_fame/HallOfFameRecordDto.java:9]: I would follow the convention we started using and include `ResponseDto` somewhere in the name

File: polymorphia-backend/src/main/resources/db/migration/V5__add_schema_for_grades.sql [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+    comment           text
+);
+
+ALTER TABLE grades
+    ADD FOREIGN KEY (gradable_event_id) REFERENCES gradable_events (id);
+
+ALTER TABLE grades
+    ADD FOREIGN KEY (animal_id) REFERENCES animals (id);
+
+CREATE TABLE criteria_grades
```
Comment (ustankie) [inline polymorphia-backend/src/main/resources/db/migration/V5__add_schema_for_grades.sql:34]: When changing the schema, please implement also the model in backend, so that everything is compatible. Now, if I wanted to create a migration model -> db (which is a common practice when chaning the backend code), these tables would be deleted as they do not exist in model

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/hall_of_fame/HallOfFameService.java [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+        Map<Long, Map<String, String>> result = new HashMap<>();
+        for (StudentScoreDetail details : detailsList) {
+            result.computeIfAbsent(details.getAnimalId(), id -> new HashMap<>())
+                    .put(details.getEventSectionName(), NumberFormatter.format(details.getRawXp()));
+        }
+        return result;
+    }
+
+    private  List<HallOfFameRecordDto> getDtoList(
+            Stream<HallOfFame> stream, Map<Long, Map<String, String>> detailsByAnimalId
```
Comment (ustankie) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/hall_of_fame/HallOfFameService.java:75]: I think it's nicer to pass `HallOfFame` here instead of a stream

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java (hunk 21 above)
Comment (ustankie) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/HallOfFameMapper.java:22]: Thanks, it's much better now!

**Review by ustankie — CHANGES_REQUESTED** [review]
I appreciate the refactor that was made, it is now much more readable, thanks! I think some tests should be added to fully complete this feature

**Review by kligarski — COMMENTED** [review]
Great job, you did really well with all complex SQL queries, annoying mappings between names and edge cases.

Note that I did not check the hibernate model, I can do that in separate review if you want.

One thing to consider is that, as far as I know, views do not speed-up any queries. It will probably be faster than our previous approach without Hibernate overhead but not faster than regular SQL query. I think that I've seen your comment somewhere about potentially using materialized views if this solution won't perform fast enough - this seems like a reasonable approach.

Another thing that we will need to handle is updating xp bonus in items, but *this is a story for another PR*.

File: polymorphia-frontend/app/(logged-in)/hall-of-fame/HallOfFameService.ts [hunk 27]
```diff
@@ -99,56 +14,46 @@ const HallOfFameService = {
...
+    const response = await fetch(
+        `${API_HOST}/hall-of-fame`,
+        {
+          method: "POST",
+          credentials: "include",
+          headers: {
+            "Content-Type": "application/json",
+          },
+          body: JSON.stringify({
+            courseId: 1,
```
[the remark this answers is not in the document]
Comment (kligarski) [reply polymorphia-frontend/app/(logged-in)/hall-of-fame/HallOfFameService.ts:35]: Okay, I realized now that #100 depends on this PR.

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/StudentDetailsMapper.java [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+    public StudentDetailsResponseDto hallOfFameToStudentDetailsResponseDto(HallOfFame hallOfFame) {
+        StudentDetailsWithoutNameResponseDto studentDetailsWithoutName =  new StudentDetailsWithoutNameResponseDto(
+                hallOfFame.getAnimalName(),
+                hallOfFame.getEvolutionStage(),
+                hallOfFame.getGroupName(),
+                hallOfFame.getImageUrl(),
+                hallOfFame.getPosition()
+        );
+
+        return accessAuthorizer.hasAnyRole(List.of(UserType.COORDINATOR, UserType.INSTRUCTOR)) ?
```
[the remark this answers is not in the document]
Comment (kligarski) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/mapper/StudentDetailsMapper.java:28]: Okay, I realized now that #100 depends on this PR.

File: polymorphia-frontend/app/(logged-in)/hall-of-fame/HallOfFameService.ts [hunk 29]
```diff
@@ -99,56 +14,46 @@ const HallOfFameService = {
...
-    let filteredData = [...ALL_DATA];
-
-    if (groups && !groups?.includes("all")) {
-      filteredData = filteredData.filter((item) =>
-        groups.includes(item.userDetails.group)
-      );
+    if (!sortBy || !sortOrder) {
+      sortBy = "total";
+      sortOrder = "desc";
     }
```
[the remark this answers is not in the document]
Comment (LisieckaK00) [reply polymorphia-frontend/app/(logged-in)/hall-of-fame/HallOfFameService.ts]: I needed to add the default value for `sortOrder` in one more place, as for some reason there appeared `undefined`.

File: polymorphia-frontend/app/(logged-in)/hall-of-fame/HallOfFameService.ts [hunk 30]
```diff
@@ -99,56 +14,46 @@ const HallOfFameService = {
...
 
-    if (searchTerm && searchTerm.trim() !== "") {
-      const lowerSearch = searchTerm.toLowerCase();
-      filteredData = filteredData.filter((item) =>
-        item.userDetails.studentName.toLowerCase().includes(lowerSearch)
-      );
+    if (sortBy === "name"){
+      sortBy = "animalName";
     }
```
[the remark this answers is not in the document]
Comment (LisieckaK00) [reply polymorphia-frontend/app/(logged-in)/hall-of-fame/HallOfFameService.ts]: Thank you for the guidance, had no idea before where to look for it. I didn't find any place more that would need the change, all seems to be working. Hoping I didn't miss anything important :P

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/model/hall_of_fame/SearchBy.java [hunk 31]
```diff
@@ -0,0 +1,28 @@
+package com.agh.polymorphia_backend.model.hall_of_fame;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+
+import static com.agh.polymorphia_backend.model.hall_of_fame.HallOfFame.FIELD_ANIMAL_NAME;
+import static com.agh.polymorphia_backend.model.hall_of_fame.HallOfFame.FIELD_STUDENT_NAME;
+
+public enum SearchBy {
+    STUDENT_NAME,
+    ANIMAL_NAME;
```
[the remark this answers is not in the document]
Comment (kligarski) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/model/hall_of_fame/SearchBy.java:10]: Okay

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/request/HallOfFameRequestDto.java [hunk 32]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.util.List;
+
+public record HallOfFameRequestDto (
+        @NotNull Long courseId,
+        @NotNull Integer page,
+        @NotNull Integer size,
+        @NotNull String searchTerm,
+        @NotNull SearchBy searchBy,
+        @NotBlank String sortBy,
```
[the remark this answers is not in the document]
Comment (kligarski) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/dto/request/HallOfFameRequestDto.java:19]: Thank you for the explanation.

**Review by ustankie — APPROVED** [review]

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/model/course/reward/assigned/AssignedReward.java [hunk 33]
```diff
@@ -0,0 +1,47 @@
...
+
+    @NotNull
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "reward_id")
+    @ToString.Exclude
+    @JsonIgnore
+    private Reward reward;
+
+    @NotNull
+    private LocalDateTime receivedDate;
```
Comment (ustankie) [inline polymorphia-backend/src/main/java/com/agh/polymorphia_backend/model/course/reward/assigned/AssignedReward.java:41]: Not sure, maybe `ZonedDateTime` would be safer?

File: polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/hall_of_fame/HallOfFameSortSpecResolver.java [hunk 34]
```diff
@@ -0,0 +1,31 @@
...
+
+    public HallOfFameSortSpec resolve(HallOfFameRequestDto requestDto) {
+        String sortBy = requestDto.sortBy();
+
+        Optional<String> overviewDbField = OverviewField.getDbFieldFromKey(sortBy);
+        if (overviewDbField.isPresent()){
+            return new OverviewFieldSort(overviewDbField.get());
+        }
+
+        boolean eventSectionExists = eventSectionRepository.existsByCourseIdAndName(requestDto.courseId(), sortBy);
```
[the remark this answers is not in the document]
Comment (kligarski) [reply polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/hall_of_fame/HallOfFameSortSpecResolver.java:24]: As discussed, we will handle this in follow-up PR.
